sdk/python/feast/registry.py

```diff
@@ -537,3 +542,80 @@ def _write_registry(self, registry_proto: RegistryProto):
         file_obj.seek(0)
         blob.upload_from_file(file_obj)
         return
+
+
+class S3RegistryStore(RegistryStore):
+    def __init__(self, uri: str):
+        self._uri = urlparse(uri)
+        self._bucket = self._uri.hostname
+        self._key = self._uri.path.lstrip("/")
+        try:
+            import boto3
+        except ImportError as e:
+            from feast.errors import FeastExtrasDependencyImportError
+
+            raise FeastExtrasDependencyImportError("aws", str(e))
+        self.s3 = boto3.resource(
+            "s3", endpoint_url=os.environ.get("FEAST_S3_ENDPOINT_URL")
+        )
+        return
+
+    def get_registry_proto(self):
+        file_obj = TemporaryFile()
+        registry_proto = RegistryProto()
+        try:
+            import botocore
+        except ImportError as e:
+            from feast.errors import FeastExtrasDependencyImportError
+
+            raise FeastExtrasDependencyImportError("aws", str(e))
+        try:
+            bucket = self.s3.Bucket(self._bucket)
+            self.s3.meta.client.head_bucket(Bucket=bucket.name)
+        except botocore.client.ClientError as e:
+            # If a client error is thrown, then check that it was a 404 error.
+            # If it was a 404 error, then the bucket does not exist.
+            error_code = int(e.response["Error"]["Code"])
+            if error_code == 404:
+                raise FeatureBucketNotExist(self._bucket)
+            else:
+                raise FeatureBucketForbiddenAccess(self._bucket)
+
+        try:
+            obj = bucket.Object(self._key)
+            obj.download_fileobj(file_obj)
+            file_obj.seek(0)
+            registry_proto.ParseFromString(file_obj.read())
+            return registry_proto
+        except botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "404":
+                raise FileNotFoundError(
+                    f'Registry not found at path "{self._uri.geturl()}". Have you run "feast apply"?'
+                )
+            else:
+                raise FileNotFoundError(
+                    f'Registry is not able to locate data under path "{self._uri.geturl()}" with [original error]: {e.response}'
+                )
```

Comment:
Can you explain what's the difference between "Registry not found at path" and "Registry is not able to locate data under path"? Should we just merge these two and raise the 1st exception?

Also, let's do `raise FileNotFoundError(...) from e` so that the original exception is maintained with its own stack trace. Then you won't have to pass `[original error]` here.

Reply:
It could be possible to get 403 error or even 500 from AWS, I think it does make sense to re-raise exception in case  in the else branch

Reply:
the same could be applied to :582 line similarly.

Reply:
The issue isn't that AWS could return other error code. It's that messages "Registry not found at path" and "Registry is not able to locate data under path" are basically telling the same thing. Maybe instead of "Registry is not able to locate data under path" we can say "Error while trying to locate Registry at path"?

Reply:
done